### ISSUE TITLE
Bug: [no-misused-promises] Fails to check unions that contain misused promises.

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8494]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXyjDBAAcMBnANRy2AGEoIIAjQgawAoA3G%2Bxl9gC54aUIiyoQweAB94HDgEp4AXgB88HrUWKA3AFgAUKEiwEYPOQzwAtlACezEAAUYOG1nIgGTVmDbCoiDiktJyCsrq8K7uniAAPFrAaopGhMRkVLw%2BAv4cdo4ubh5eOX5sekA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare function acceptsVoidCallback(voidCallback: undefined | (() => void));
declare const maybePromiseCallback: undefined | (() => Promise<void>)
acceptsVoidCallback(maybePromiseCallback);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-misused-promises": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that the 3rd line should report "Promise returned in function argument where a void return was expected"

### Actual Result

There was no error on the 3rd line.

### Additional Info

This continues the conversation from https://github.com/typescript-eslint/typescript-eslint/issues/5744#issuecomment-1947717802.

The rule does already check arguments that have multiple call signatures, but it doesn't check whether it is a union of something that has call signatures and something that doesn't. 

We should also try to ensure that complex test cases around (possibly multiple) spread args behave appropriately, including yucky things like 
```ts
declare function acceptsVoidCallbacks(...voidCallbacks: Array<undefined | (() => void)>);
declare const maybePromiseCallback: undefined | (() => Promise<void>)
declare const voidCallback: () => void;
acceptsVoidCallbacks(voidCallback, undefined, ...[voidCallback, voidCallback], ...[maybePromiseCallback], ...[voidCallback]);
```
